### DISCUSSION TITLE
Debian patches

### DISCUSSION THREAD

**Discussion by Elliria** [autokey/autokey#880] [Maintenance goals]
Here are some [Debian patches](https://udd.debian.org/patches.cgi?src=autokey&version=0.95.10-2.1) that have been applied on Debian, but that never got forwarded to us. They may be helpful for our efforts to clean up and update the code.

**Comment by ineuw** [comment]
Are these patches applicable to 96.0?

**Reply by josephj11** [reply]
@Elliria These are very old and were all incorporated in 0.95.10 and as such, became part  of 0.96.0.

These changes were (partially) made to get Debian's blessings for 0.95.10 - which they did subsequently package. This is why all Debian-based distros offer 0.95.10 as the current version.

Since then, there were new packaging problems for 0.96.0 which prevented it from being included in Debian. Elliria has recently fixed these problems, but Debian hasn't noticed yet.

It references three patches.

The first one was done by our Debian packager and just updates a dependency. This should be updated in our current dependencies.

It's in ./lib/autokey/gtkui/notifier.py
Doesn't appear to be relevant in the Qt version.

The other two were done by us (Sam).

The second one is done because our desktop files include that parameter with all those arguments and a few more. This was filed as issue #370 and implemented in PR #480 .

The third one is done  because that option is in the autoKey-qt man page for 0.96.0...

**Reply by ineuw** [reply]
Would it be possible add a revision number and offer a single version based on the current code?

**Reply by josephj11** [reply]
@ineuw Not sure what you are asking. These particular changes are all long done and incorported.

Almost sounds like this comment was meant for the other thread that we both just replied to.

We can create and release 0.96.1 whenever we want to (as a maintenance release). There's a separate discussion #770 for what should be included in that and when it should be released.

**Reply by Elliria** [reply]
Nope. My mistake. They've all been done already.

**Reply by Elliria** [reply]
You can [clone](https://github.com/autokey/autokey/wiki/Running-Tests#trying-out-a-clone-of-autokey) the current code at any time, @ineuw.

**Comment by ineuw** [comment]
In your opinions, what is the most important issue that should be addressed, before releasing a maintenance version?

**Reply by josephj11** [reply]
This  discussion should probably be elsewhere. There's a whole 0.96.1 discussion thread (warning, it's really long already) which might be more appropriate. Or maybe a new discussion about AutoKey policy and working assumptions, etc.

 OFC, that changes for each release. The most important things would be regressions - things that the current release unintentionally broke.

Since it's a maintenance release, it would usually not include any new features unless they're very minor and non-breaking - don't cause any backward compatibility issues. Exceptions could be made if something new was critical to keep the application working when something external has changed...

Since it's still a new release and entails a lot of community effort, we need to make sure that it has enough in it to justify doing it at all.

For 0.96.1, my pet peeve is #600 - the Wayland detector. Other than that, I don't have specifics. We will have to look at what's ready to go and what seems important to most of the team and users. When we're almost ready we should post on our forums and see if we get any user feedback.

I think #600 is enough by itself to justify a release. People have already spent a lot of time trying to figure out why everything seems broken when a simple error message could have made that clear immediately.

I would also include the changes to the CLI options that Elliria recently made. They don't break anything and just make life a little simpler.

**Reply by Elliria** [reply]
> In your opinions, what is the most important issue that should be addressed, before releasing a maintenance version?

I would say figuring out how to release a maintenance version would have to be at the top of my list. For example, which conflicts would need to be resolved, which branch(es) would need to be merged, which branch(es) would need to be "frozen", what kind of testing would need to be done, etc.

**Reply by josephj11** [reply]
We definitely need to get that down. So at least whatever problems doing it presents will be the same ones everytime and we'll know how to deal with them.


Conceptually, minor and major releases should be done almost identically (git can't tell the difference). But if there are things in develop that shouldn't go into the next beta (quite likely to occur), then we'll have to learn about cherry picking...

We also have to have a policy/procedure for what to do when a fix is applied to the beta. How does it get into develop as well?

When the beta is done, it "should" replace the master and we should create a release tag that triggers the release processs.

What all this actually means and how to do it will need more experimentation and documentation. This should end up being a relatively smooth process, but we have a ways to go on that.

This is way more complicated  than I expected it to be. The upside is that the more you learn of this rocket science, the more marketable your skills are.

Given the complexity involved, I would expect that competent project/git managers probably command significant salaries/fees.

**Reply by ineuw** [reply]
Can we consider a fundraising project like **gofundme** for tech projects? I am willing to write the story on how an Autokey user suffers, just by reading these posts. 

Based on the open issues: 
1. Many of them are using an esoteric environment with which we are not equipped to deal with.
2. The posters don't care about continuing or closing the issue as a common courtesy, whatever the outcome was.
3. After six months, we should close them as **unresolved**. The speed of upgrades and improvements that are being introduced on a daily basis correct many old issues, and (naturally), create new ones.

We should focus on fundamental issues surrounding Autokey. By the surrounding, I mean updates to libraries required by Autokey. Historic issues with GTK support, etc.

I would like to see input from all users, not just the ones who report issues. I regret not being able to follow discussions of problems and their resolutions elsewhere.

**Reply by josephj11** [reply]
TL;DR: It's more complicated than it sounds.

Issues:

Much as I like money and woud love to be paid, what difference would having (presumably small) amounts of money make?

Theoretically, I have several ways to accept payments now, but those funds would come to me and be mine. It would be taxable income. Obviously, this wouldn't be fair to the other contributors to the project.

If you want the funds to come to "AutoKey" where would they go?

Someone would have to set up an organization like a 501C3 corporation which costs money to do and it would involve recurring work and expenses for license renewals, filing tax returns, etc.

It's a nightmare unless someone decides to donate thousands of dollars.

Assuming we did all that, we would still have to pay a substantial amount to a competent programmer to do anything. That would require contracts, project management, etc.

I would love to be wrong about this!

For my one-person project, I set up a donation link years ago on sourceforge. No one has ever used it in almost two decades. I can't even get forum comments or reviews let alone money. If I did, it would be personal income which is no big deal to account for.

A group project is a whole other deal.

**Reply by Elliria** [reply]
Since @josephj11 hasn't had a nibble in nearly twenty years, it's unlikely to go anywhere, but you can try, @ineuw. I'd say most of us who have contributed and are contributing to the project have been and are in it for the love of the program, but it couldn't hurt to attempt to lure in knowledgeable programmers with cold, hard cash. Meanwhile, we'll mosey along and slowly figure this thing out.

**Reply by josephj11** [reply]
@Elliria We have had a number of users offer to donate to the project, but it never got to the point of mentioning an amount.

The first problem is that there's nowhere to transfer any donated funds to (from a donation service or directly from an individual). The IRS  or its equlivalent in another country is going to want to collect taxes on it or get proof that the recipient is tax exempt. These days, there are also AML (anti money laundering) concerns that have to be addressed.

In the simplest case, if someone sent me money directly, I would be liable for any taxes on it. If I then wanted to split the proceeds with the rest of you, I would have to calclulate the net after taxes first and then each of you would have to pay taxes (again) on your cut.

The only people likely to benefit from such transactions would be our tax accountants.

Failure to account for the income could open one up to legal and financial difficulties.

Again, if the amount(s) donated is large, then such machinations might be justified.

If we all had Lightning wallets, we could "zap" each other Sats (tiny fractions of a Bitcoin). That would allow us to remain pseudonmyous, but that still wouldn't address accounting for it.

Most of you probably wouldn't want to reveal a conventional bank account or other personal information so I had somewhere to send your cut to.

**Reply by Elliria** [reply]
I number among those who wouldn't. In the end, a project like ours will benefit the most from those who love it, I think.

**Comment by josephj11** [comment]
This stuff is just part of the FOSS ecosystem. The only ways around it are to get really big, offer support contracts/consulting, or get large paying sponsors.

Until I had my own project, I didn't even think about contacting the projects of things I downloaded. Now, I sometimes go out of my way to do so because I found out how lonely it gets if you're not a super star.

**Reply by ineuw** [reply]
@josephj11 - An excellent assessment. I remember discussing this in the past with you. [and this article agrees with your experience and assessement](https://stackoverflow.blog/2021/01/07/open-source-has-a-funding-problem/) But, I am an optimist. We should assess the time realistically. You should determine what is the most important corrections need to be done.

Generally, I have more time than either you or @Elliria, but I am not qualified. Also, my experiences are with using Autokey in browsers and the web. Firefox in particular, and then Chrome/Chromium and Vivaldi. I rarely use Autokey offline.

**Reply by josephj11** [reply]
Also remember, that we don't have any governance policy here. The only restraint is that AFAIK, I am the only active member who can grant elevated access  to our repo to a new developer (although I may have given that level to Sam).

Other than that, we each have our opinions as to what should be done, but a developer can work on whatever they want to. (If we gave them a hard time, they could just fork AutoKey and ignore us. Some people already do this and it's not a problem.)

We have a say in what PRs get merged, what goes into a release, and when releases come out.

I sort of run things, but only because I am tolerated by the rest of you. And I try to do whatever the team wants unless it's something I strongly disagree with (which hasn't happened yet).

You and anyone else who puts in substantial work on the project have an equal say as to what gets done. I just help make it happen and help keep it going.

Feel free to review all the open issues and point out any which you feel should have higher or lower prioritiy. (Keeping in mind that some great ideas aren't going to be done any time soon because we lack developer resources.)

If you know someone interested in supporting the project financially in a substantial way, we can certainly consider that. It would give them a lot of influence over the project.

I would probably not be interested in running a formal AutoKey organization. It sounds like quite a bit of additional work. I have no experience running organizations and I'm not sure I would be good at it. I don't think I could do it for free.

Just read that article and bookmarked it. Thanks. It's thoughtfully written.
